compiler/rustc_mir/src/transform/simplify.rs

```diff
@@ -59,12 +68,31 @@ impl<'tcx> MirPass<'tcx> for SimplifyCfg {
         Cow::Borrowed(&self.label)
     }
 
-    fn run_pass(&self, _tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
         debug!("SimplifyCfg({:?}) - simplifying {:?}", self.label, body.source);
-        simplify_cfg(body);
+        simplify_cfg_with_coverage(tcx, body, DroppedCoverage::SaveUnreachable);
     }
 }
 
+/// How to handle `Coverage` statements in dropped `BasicBlocks`.
+pub enum DroppedCoverage {
+    /// Allow the `Coverage` statement(s) in a dropped `BasicBlock` to be
+    /// dropped as well. (This may be because the `Coverage` statement was
+    /// already cloned.)
+    Allow,
+
+    /// Assert that there are none. The current or most recent MIR pass(es)
+    /// should not have done anything that would have resulted in an
+    /// unreachable block with a `Coverage` statement.
+    AssertNone,
```

Comment:
I'm a bit curious why you took this approach. I think I'm probably missing some context here so additional details would be greatly appreciated! 🙂 

At least from a cursory review, it seems like the following is true:
- When `-Zinstrument-coverage=off`, we never generate MIR coverage statements to begin with and we also don't care if they were made unreachable by a MIR pass. 
- When `-Zinstrument-coverage=on`, it doesn't really matter that a pass made a coverage statement unreachable as long as keep a coverage statement in `bb0` which marks the code region as unreachable.

(I suspect the context I'm missing is why my second statement is not true for some reason.)

If both of those statements *were* true, then it seems like we don't need this enum at all, we just need to know if `-Zinstrument-coverage=on` and in that case, make sure we retain the code regions as unreachable. 

What I like about that is it simplifies the calling code so they don't have to be aware of code coverage at all, they just have to pass `tcx` down. 

Thoughts?

Reply:
It's a very reasonable question.

I THINK this assumption is what you may have wrong:

> * When -Zinstrument-coverage=on, it doesn't really matter that a pass made a coverage statement unreachable as long as keep a coverage statement in bb0 which marks the code region as unreachable.

I tried a very simple implementation of this, and it didn't work. One reason requires special handling:

Some passes clone blocks, and use dead-block removal to remove the original statements. `simplify_cfg` has no was to know this, so when the coverage statements from the dead block are added to `bb0`, they are duplicate code regions. The cloned blocks still have the original _reachable_ blocks and coverage statements. Adding `unreachable` code regions to `bb0` breaks coverage reports, probably because the same code region is reported as both reachable and unreachable. So we need the `Allow` enum option for this case (allow the dead blocks' coverage statements to be dropped).

So I need to at least call `simplify_cfg` with an option to either `Allow` dropping the coverage, or to `Save` the coverage.

Here's why I added the `AssertNone` option, which is the default, if the caller uses the existing function signatures (plus `tcx`):

In my view, I did "simplify the calling code so they don't have to be aware of code coverage" for those passes that don't (today) do anything that would affect code coverage.

I simplified it by keeping the original functions `simplify_cfg()` and `remove_dead_blocks()` (only adding the `tcx` parameter).

The passes that call these functions are assumed to be "safe" to coverage.

But the problem is, the only way I found out that there were a few RARE edge cases that dead-block removal broke was during codegen. By that time, I have no idea what pass broke coverage during CFG simplification.

By adding the assertion and check, if a new edge case pops up that we can't predict, we'll have a lot more to go on.

Obviously, we could remove this assert, and assume that I've caught all possible edge cases; but since I got bit twice already, I don't feel like I can make that assumption. And a new pass, or change to a MIR pass, might break coverage in the future. This is obscure enough that I think we should check.